goalie/mesh_seq.py

```diff
@@ -684,92 +469,133 @@ def solutions(self):
             self._create_solutions()
         return self._solutions
 
+    def _reinitialise_fields(self, initial_conditions):
+        """
+        Reinitialise fields and assign initial conditions on the given subinterval.
+
+        :arg initial_conditions: the initial conditions to assign to lagged solutions
+        :type initial_conditions: :class:`dict` with :class:`str` keys and
+            :class:`firedrake.function.Function` values
+        """
+        for field, initial_condition in initial_conditions.items():
+            fs = initial_condition.function_space()
+            self.fields[field] = (
+                firedrake.Function(fs, name=field),
+                firedrake.Function(fs, name=f"{field}_old").assign(initial_condition),
+            )
+
     @PETSc.Log.EventDecorator()
-    def solve_forward(self, solver_kwargs={}):
+    def _solve_forward(self, update_solutions=True, solver_kwargs={}):
         r"""
-        Solve a forward problem on a sequence of subintervals.
-
-        A dictionary of solution fields is computed - see :class:`~.ForwardSolutionData`
-        for more details.
+        Solve a forward problem on a sequence of subintervals. Yields the final solution
+        on each subinterval.
 
+        :kwarg update_solutions: if ``True``, updates the solution data
+        :type update_solutions: :class:`bool`
         :kwarg solver_kwargs: parameters for the forward solver
         :type solver_kwargs: :class:`dict` whose keys are :class:`str`\s and whose values
             may take various types
         :returns: the solution data of the forward solves
         :rtype: :class:`~.ForwardSolutionData`
         """
+        # TODO docstring tag for yield?
         num_subintervals = len(self)
-        P = self.time_partition
-        solver = self.solver
+        tp = self.time_partition
 
         # Reinitialise the solution data object
         self._create_solutions()
+        solutions = self.solutions.extract(layout="field")
```

Comment:
Actually solutions should be reinitialised only if `update_solutions=True` I guess

Reply:
Done in c86222b340fcdd039c178b1b1adc79544fc0deb1